### ISSUE TITLE
"too many open files" in static sync

### ISSUE THREAD

**Issue by bep** [spf13/hugo#1872]
I got a "too many open files" error on OS X, probably related to doing lots of /static syncing. Suspect file opened, but not closed.

**Comment by bep** [comment]
Not related to #1870

**Comment by bep** [comment]
> It would be nice if this was run automatically for OSX users, I am not sure of the purpose of the command currently.

There is (or was) a tweek in Hugo to set this limit on OS X. Not sure if this is related.

**Comment by mattstratton** [comment]
We still are seeing this issue on OS X. We have attempted to use `hugo check ulimit` but that doesn't solve for it the only fix is to run 
``` $  sudo sysctl -w kern.maxfiles=65536
 $  sudo sysctl -w kern.maxfilesperproc=65536
 $  ulimit -n 65536 65536```

This doesn't persist after a reboot, so it's a bit of a problem. Especially since we have dozens of contributors. 

See https://github.com/devopsdays/devopsdays-web to repro.

**Comment by moorereason** [comment]
Persistent sysctl settings are usually in /etc/sysctl.conf.
```
echo kern.maxfiles=65536 >> /etc/sysctl.conf
echo kern.maxfilesperproc=65536 >> /etc/sysctl.conf
```

**Comment by mattstratton** [comment]
Regardless, it's still a problem that we have to give all of our contributors persistent sysctl settings in order to contribute content :(

This is definitely an improved workaround, but I am still curious as to why `hugo check ulimit` doesn't work?

**Comment by mattstratton** [comment]
For some background, there are 4,696 files in the `/static` directory (and subdirectories). Yes, I know that's horrid, but it was an artifact of migration (we had so much legacy stuff to migrate that the expedient workaround was to take the output HTML from the former site generator and dump it into `/static`)

**Comment by moorereason** [comment]
So what are the default values for `kern.maxfiles` and `kern.maxfilesperproc` on OSX?

Hugo [attempts to increase the resources limits](https://github.com/spf13/hugo/blob/2c5e4f7640e71d2a193a74e6c41109ec40bc0222/commands/limit_darwin.go), but you may have to run Hugo with elevated privileges for that to actually work.

The only "real" solution I can think of from Hugo's perspective would be to limit the number of open file descriptors to keep them under the process limits, but that could get very complicated.

**Comment by mattstratton** [comment]
These are the defaults:

```
kern.maxfiles: 12288
kern.maxfilesperproc: 10240
```

**Comment by yaleman** [comment]
I'm curious to know why hugo wants to keep so many files open at once? Surely it shouldn't have to when generating static content? I understand that it might be an issue with `hugo serve` without being able to subscribe to filesystem updates.

Strangely it works when I run it a couple of times in a row. I have 11045 pages, and it creates 5773 each of tags and categories.

**Comment by moorereason** [comment]
> I'm curious to know why hugo wants to keep so many files open at once?

Parallelism.

**Comment by yaleman** [comment]
From what I can tell it's when I have many categories/tags. Drop the amount back and it stops being an issue. Noone's got ~10k cores, so "paralleism" is probably not the best answer at that point.

**Comment by bep** [comment]
This issue has been automatically marked as stale because it has not been commented on for at least four months.

The resources of the Hugo team are limited, and so we are asking for your help.

If this is a **bug** and you can still reproduce this error on the <code>master</code> branch, please reply with all of the information you have about it in order to keep the issue open.

If this is a **feature request**, and you feel that it is still valuable, please open a proposal at https://discuss.gohugo.io/.

This issue will automatically be closed in four months if no further activity occurs. Thank you for all your contributions.

**Comment by bep** [comment]
**Note/Update:** This issue is marked as stale, and I may have said something earlier about "opening a thread on the discussion forum". Please don't.

If this is a **bug** and you can still reproduce this error on the latest <code>release</code> or the <code>master</code> branch, please reply with all of the information you have about it in order to keep the issue open.

If this is a **feature request**, and you feel that it is still relevant and valuable, please tell us why.

**Comment by jamiedumont** [comment]
@bep I've recently experienced this issue on a new project due to a large number of folders/files in `/static`. I'm not sure of the threshold (number of files) that prompts the error, but it doesn't seem excessive (~2000 files across a folder tree causes it in my case).

I only experience the issue running `hugo server`. The build (`hugo`) is of course fine.

So far I have tried all the suggestions relating to `ulimit` in this issue, and also other suggestions I've found elsewhere (updating homebrew and updating hugo), without success.

Of course, an alternative would be to find a different work flow for development, but this issue seems persistent, so is perhaps worth resolving.
